### ISSUE TITLE
Add more toggles for Ace Editor

### ISSUE THREAD

**Issue by mclindy** [Atheos/Atheos#245]
**Is your feature request related to a problem? Please describe.**
Not related to a problem. The only "problem" is that I can get most of what I want by manually editing code, but that's more like cutting out parts that I'll later want to put back into use.

I need the ability to set the following as default, then include toggles in the Editor Setting panel to turn them on if desired:
enableBasicAutocompletion: false
enableSnippets: false
enableLiveAutocompletion: false

> [EDIT] Actually, after googling and reading up on ace.c9.io, I think I need **behavioursEnabled: false** (sic). My "test" of whether my hack was successful was observing tag completion in an HTML file, and setting autocompletion values to false in the /components/editor/init.js file didn't stop tag completion. It probably _did_ stop other kinds of autocompletion, but I wasn't looking for that type of autocompletion at the time. I realize now that I am _not_ getting a pop-up of code completion suggestions as I type, so setting these values as the default in init.js is most likely working as designed. It appears that I did not understand what they did. I'm editing this to stop you spending time on this and having me say "it doesn't work" when in fact what I want is to also disable HTML tag completion (auto pairing?). [slaps head]

> [EDIT 2] I hate to admit it, but merely adding **behavioursEnabled: false** to /components/editor/init.js after approx line 116 enableLiveAutocompletion gives me the results I was looking for. So on the one hand, my immediate problem is solved. On the other hand, it would still be wonderful to be able to toggle these on and off as requested. I know you're on it--I just want to clarify. Thank you for your patience. I'm leaving this post up in case my errors-in-haste are helpful to anyone in the future. Beware the spelling: it's behavi**ou**rs with an OU, as in British spelling, not the American spelling! [slaps head even harder]

**Describe the solution you'd like**
~~First, pease tell me where can I manually edit a file to set defaults of autocomplete, etc. _/components/editor/init.js_ looks promising but changing the values there doesn't seem to help. I know I'm missing something, but not sure where to look.~~[Resolved, see edits above]

Eventually this could be done with a new administrative panel that would be in the settings in the right sidebar.

I imagine a solution along these lines;

- Admin user an create types of users, e.g. student, teacher, etc.
- Admin can then assign users to a group. That would probably require a new property (e.g. userType in the create method in _class.user.php_).
- Configuration settings could be adjusted for groups. This could be done with a new settings modal, or it could be done by adding an "Apply to: [select group]" line in an existing settings modal.
- Admin could enable the ability of group members to toggle settings that are otherwise hidden (?)

Hopefully you get the idea. Now that I've written this much I wish that I would have chosen a different name for this post. "Add more toggles for Ace Editor" was where I started, but I've ended up going a bit past that. :)

**Additional context**
I teach web design and programming classes at a high school in Oregon, USA. We use a text-editor IDE from day one. In the past that has been Notepad++. Covid made us switch to Chromebooks, and I ended up using CodeTasty. I switched to Atheos because CodeTasty's server(s) have been unreliable, which has wreaked havoc on new coders' daily experience. Further, it has always made me nervous to have a service-in-the-middle where I have no idea about their security! For the record, CodeTasty has been great, but Atheos is a much better fit for me and my students.

I find that for brand-new students it is very effective to disable autocomplete and code suggestions. This forces them to think of "starting" and "stopping" parts of scripts/code, and it requires them to memorize some basic terminology (e.g. HTML elements and programming terms, such as function() ). Doing this manually for a while teaches them how to "think in code." Eventually you turn the autocomplete on and you're more efficient, but you are a better debugger for learning how to "think" first, IMO.

Anyhow, I'd love the ability to have this off by default, then be able to turn this (and other features) on.

I found Atheos after looking for solutions when CodeTasty started having server issues. I looked into CodeAnyhwere, but it's expensive and honestly a bit too involved for brand new, young programmers.

I currently have about 150 students using Atheos and it's a great success so far. I am going to use it for a coding camp with about 75 middle schoolers in March.

Adding these features will be a big help for me, but I think would also help this program be picked up by educators everywhere. It's pretty amazing as it is, but adding a few (relatively) simple user management features could help revolutionize how schools teach programming!

I'll help any way I can.

Thanks,
Greg

**Comment by HLSiira** [comment]
I can work on this this weekend and put something together; this one should be pretty simple to add, assuming that I didn't leave my install of Atheos in a messy state since I was last working on it lol. This is really cool to read about though, I hope they're having fun using Atheos.

**Comment by mclindy** [comment]
Thanks! I'm running a demo that I can hack on while the students are working with the working site.

Here's what I've done so far, though your interpretation will be better:
![settingsMod](https://user-images.githubusercontent.com/16168011/213208844-6aa7b7c4-0f82-45f2-a29e-25bb2ec2543c.jpg)
That involved adding lines of code to /components/editor/init.js at around line 395 (right after Set Font Size.

FWIW, I tried to manually set the values under applySettings() in lines 114 - 116 to false, but that doesn't turn any features off. I also edited /components/settings/panels/editor.php and added lines at around line 46. While the Settings modal seems to behave well and pass values to editor.php I am unable to affect the operation of Ace Editor. 

Welp, I'm off to a faculty meeting.

I look forward to using your solution and I have no stake or interest in perpetuating my hacks. :)

Thanks,
Greg

BTW yes, the students are raving about it. It's soooo much better on Chromebooks than anything else I know about! In our lab we use proper desktops and the students are switching to Atheos over CodeTasty or Notepad++ as soon as they try it.

**Comment by mclindy** [comment]
One more feature request: drag-n-drop to move files and/or folders within the directory tree. Sometimes, as they're learning, students create files and/or directories in the wrong location. At the moment, the only way to "move" a file or directory is to create a new one in the desired location, then copy & paste the text contents as needed, etc. If drag-n-drop is too cumbersome, then a **right-click->move** in the directory tree would also solve the problem.

Thanks. I realize you're busy with life, work, etc. I also realize that perhaps this is something you can't get to, and I respect that. I didn't post it in another request because it seems that I have ended up asking for a few features that specifically target new, low-skilled coders (e.g. students) and I feel that at some point my requests cross a line and go beyond what you intended to develop and support.

Thanks,
Greg

**Comment by mclindy** [comment]
Re: Drag N Drop

I have never been able to drag-n-drop. This goes back to uploading files--I'm unable to drag N drop onto the pop-up, but using the file search dialogue works well. I'll be my issue with drag N drop is limited to my installation setup since it was not mentioned in [issue 243](https://github.com/Atheos/Atheos/issues/243). _That_ issue is concerned with the ability to upload _at all_; it does not mention anything about the ability to drag N drop. I think someone else would've posted this as a problem if indeed it was an Atheos problem. I'm beginning to think it's a configuration problem, but I don't know what that could be. Perhaps it's my PHP version? Can anyone confirm? I think I'm wasting my searching through code, especially if I'm the only one with this problem.

Basic config: 
- Rocky 8.7
- Apache 24.37
- PHP 8.0.27

**Comment by coca348** [comment]
See #255  for drag n drop on transfer's dialog !

**Comment by mclindy** [comment]
Thank you, coca348! I can confirm that your solution works! I can now use drag n drop to upload files.

**Comment by mclindy** [comment]
Thanks again, coca348. Based on your solution I am able to focus a little more. Perhaps this is a new issue and I should post it as such, but I hesitate to clutter up the main page if this is still just a problem with my own setup:

- I _can_ drag n drop files into an _open_ (i.e. expanded by clicking the plus sign) directory.
- I _can't_ drag n drop files onto a _closed_ directory.
- - Thus, I can't create a new (i.e. empty) directory and then drag files into it.

I can kind of get around this by:
- From the root directory, create a new directory named _test_
- Inside /test, make a new file called test.txt
- Double-click the new directory so that it opens and the internal file is visible
- Drag a file from outside of the directory into the directory
- Delete test.txt

Can anyone else create a new, empty directory, then drag files into it without the step I've described above? I think that many/all of you use Atheos for daily development, that's why I think that perhaps this is a problem that only I am having. Or perhaps it's just that it's rare for a user to create a directory for the sole purpose of moving other content into it. Typically, I'd create a directory and then create it's index.html, etc. 

Thanks. So close now!

**TLDR: Want to be able to create a new directory then be able to drag n drop into that empty directory.**

**Comment by HLSiira** [comment]
Being able to drag and drop files into a closed folder is one of my biggest frustrations with Atheos right now. The mechanism by which I built that works so well but is incredibly limited and I haven't found a way to really add anything to it.

That aside, this issue has gotten a little busy for me to fully follow right now, but I managed to make working on Atheos a permanent part of my schedule so this will be implemented / fixed, including the UX settings you asked about.

**Comment by mclindy** [comment]
Thank you. I love it, BTW. Trying to help make it better. Sometimes when people do that (myself included) it can sound like they're not grateful and/or enjoying what they have. So if I haven't already, I'd like to say thanks a ton for what I already have. It's already better than anything I had found for a classroom in almost ten years. I can't wait to start my new outreach program at the middle school in two weeks. I'll be using different installations for each middle school, e.g. school1.mydomain.com, school2, etc. That will keep kiddo accounts separated/isolated in case separated in case I mess something up! Again, super excited. Atheos really has made my implementation so much better than it otherwise would have been. Peace.
